habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1289,16 +1341,310 @@ def step(
+            wait_duration,
+            *args,
+            **kwargs,
+        )
+
+    def _check_stop(self, xyt_waypoint) -> bool:
+        """Use a heuristic to determine if an action qualifies as a stop action"""
+        linear_diff = np.linalg.norm([xyt_waypoint[:2]])
+        angular_diff = abs(xyt_waypoint[2])
+
+        lin_check = linear_diff < self._min_abs_lin_diff
+        ang_check = angular_diff < self._min_abs_ang_diff
+        return lin_check and ang_check
+
+    def _preprocess_action(self, xyt_waypoint):
+        """Perform scaling and clamping of input"""
+        # Scale
+        if self._enable_scale_convert:
+            xyt_waypoint[0] = self._scale_inputs(
+                xyt_waypoint[0],
+                [-1, 1],
+                [
+                    self._waypoint_lin_range[0],
+                    self._waypoint_lin_range[1],
+                ],
+            )
+            xyt_waypoint[1] = self._scale_inputs(
+                xyt_waypoint[1],
+                [-1, 1],
+                [
+                    self._waypoint_lin_range[0],
+                    self._waypoint_lin_range[1],
+                ],
+            )
+            xyt_waypoint[2] = self._scale_inputs(
+                xyt_waypoint[2],
+                [-1, 1],
+                [
+                    self._waypoint_ang_range[0],
+                    self._waypoint_ang_range[1],
+                ],
+            )
+
+        # Clamp
+        xyt_waypoint_clamped = np.array(
+            [
+                np.clip(xyt_waypoint[0], *self._waypoint_lin_range),
+                np.clip(xyt_waypoint[1], *self._waypoint_lin_range),
+                np.clip(xyt_waypoint[2], *self._waypoint_ang_range),
+            ]
+        )
+
+        # Convert deg to rad
+        if self._yaw_input_in_degrees:
+            xyt_waypoint_clamped[2] = np.deg2rad(xyt_waypoint_clamped[2])
+
+        return xyt_waypoint_clamped
+
+    def _step_rel_waypoint(
+        self, xyt_waypoint, max_wait_duration, *args, **kwargs
+    ):
+        """Use the waypoint-to-velocity controller to navigate to the waypoint"""
+
+        # Initialize control loop
+        xyt_init = self._agent_state_to_xyt(self._sim.get_agent_state())
+        self.w2v_controller.set_goal(
+            xyt_waypoint, start=xyt_init, relative=True
+        )
+
+        xyt = xyt_init.copy()
+        dt = self._time_step
+
+        # Forward simulate
+        for _t in np.arange(0, max(max_wait_duration / dt, 1)) * dt:
+            # Query velocity controller for control input
+            linear_velocity, angular_velocity = self.w2v_controller.forward(
+                xyt
+            )
+
+            # Apply action and step simulation
+            next_agent_state = self._apply_velocity_action(
+                linear_velocity, angular_velocity, time_step=dt
+            )
+            xyt = self._agent_state_to_xyt(next_agent_state)
+
+            # Complete action early if commanded speed is low
```

Comment:
This condition is essentially checking if the agent has been commanded to stop by the controller. So it should be above the `self._apply_velocity_action()` function.

Reply:
It doesn't really matter if we execute the last action or not if the velocity is low.
Having it here will guarantee that we take at least one step and `next_action_state` is populated.